libsolidity/analysis/ControlFlowRevertPruner.cpp

```diff
@@ -0,0 +1,223 @@
+
+
+namespace solidity::frontend
+{
+
+namespace {
+
+/// Find the right scope for the called function: When calling a base function, we keep the most derived, but we use the called contract in case it is a library function or nullptr for a free function
+ContractDefinition const* findScopeContract(FunctionDefinition const& _function, ContractDefinition const* _callingContract)
+{
+	if (auto const* functionContract = _function.annotation().contract)
+	{
+		if (_callingContract && _callingContract->derivesFrom(*functionContract))
+			return  _callingContract;
+		else
+			return functionContract;
+	}
+
+	return nullptr;
+}
+}
+
+void ControlFlowRevertPruner::run()
+{
+	// build a lookup table for function calls / callers
+	for (auto& [pair, flow]: m_cfg.allFunctionFlows())
+		collectCalls(*pair.function, pair.contract);
+
+	findRevertStates();
+	modifyFunctionFlows();
+}
+
+FunctionDefinition const* ControlFlowRevertPruner::resolveCall(FunctionCall const& _functionCall, ContractDefinition const* _contract)
+{
+	auto const& functionType = dynamic_cast<FunctionType const&>(
+		*_functionCall.expression().annotation().type
+	);
+
+	if (!functionType.hasDeclaration())
+		return nullptr;
+
+	auto const& unresolvedFunctionDefinition =
+		dynamic_cast<FunctionDefinition const&>(functionType.declaration());
+
+	FunctionDefinition const* returnFunctionDef = &unresolvedFunctionDefinition;
+
+	if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
+	{
+		if (*memberAccess->annotation().requiredLookup == VirtualLookup::Super)
+		{
+			if (auto const typeType = dynamic_cast<TypeType const*>(memberAccess->expression().annotation().type))
+				if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+				{
+					solAssert(contractType->isSuper(), "");
+					ContractDefinition const* superContract = contractType->contractDefinition().superContract(*_contract);
+
+					returnFunctionDef = &unresolvedFunctionDefinition.resolveVirtual(
+						*_contract,
+						superContract
+					);
+				}
+		}
+		else
+		{
+			solAssert(*memberAccess->annotation().requiredLookup == VirtualLookup::Static, "");
+			returnFunctionDef = &unresolvedFunctionDefinition;
+		}
+	}
+	else if (auto const* identifier = dynamic_cast<Identifier const*>(&_functionCall.expression()))
+	{
+		solAssert(*identifier->annotation().requiredLookup == VirtualLookup::Virtual, "");
+		returnFunctionDef = &unresolvedFunctionDefinition.resolveVirtual(*_contract);
+	}
+
+	if (returnFunctionDef && returnFunctionDef->isImplemented())
+		return returnFunctionDef;
+
+	return nullptr;
+}
+
+void ControlFlowRevertPruner::findRevertStates()
+{
+	std::set<CFG::FunctionContractTuple> pending_functions = keys(m_functions);
+
+	while (!pending_functions.empty())
+	{
+		CFG::FunctionContractTuple item = *pending_functions.begin();
+		pending_functions.erase(pending_functions.begin());
+
+		auto& revertState = m_functions[item];
```

Comment:
I think it's really helpful to move the declaration of a non-const auto reference as close to its use as possible. I think it's much more important to realize that `revertState` is a non-const reference towards the end of this function than is the (potential!) additional runtime of evaluating `m_functions[item]` twice.